src/set_theory/surreal.lean

```diff
@@ -269,28 +439,200 @@ instance : has_lt surreal   := ⟨lt⟩
+    rwa [← int.neg_succ_not_pos m] }
+end
+
+lemma gmul_cancel_of_pos {m : ℤ} {x y : surreal} (hm : 0 < m) (hmxy : m • x  = m • y) : x = y :=
+begin
+    contrapose hmxy,
+    cases (@ne_iff_lt_or_gt _ _ x y).1 hmxy with hmxy' hmxy',
+    { apply _root_.ne_of_lt,
+      exact lt_of_gsmul_pos_lt hm hmxy' },
+    { apply ne_of_gt,
+      change y < x at hmxy',
+      exact lt_of_gsmul_pos_lt hm hmxy' }
+end
+
+-- TODO: move this to a better location
+lemma int.pow_right_injective {x : ℤ} (h : 2 ≤ x) : function.injective (λ (n : ℕ), x ^ n) :=
+λ n m hnm, begin
+  lift x to ℕ using (zero_le_two).trans h,
+  norm_cast at h hnm,
+  exact nat.pow_right_injective h hnm
+end
+
+/-- The surreal number `half`. -/
+def half : surreal := ⟦⟨pgame.half, pgame.numeric_half⟩⟧
+
+/-- Powers of the surreal number `half`. -/
+def pow_half (n : ℕ) : surreal := ⟦⟨pgame.pow_half n, pgame.numeric_pow_half⟩⟧
+
+@[simp] lemma pow_half_zero : pow_half 0 = 1 := rfl
+
+@[simp] lemma pow_half_one : pow_half 1 = half := rfl
+
+@[simp] theorem add_half_self_eq_one : half + half = 1 :=
+quotient.sound pgame.add_half_self_equiv_one
+
+lemma double_pow_half_succ_eq_pow_half (n : ℕ) : 2 • pow_half n.succ = pow_half n :=
+begin
+  rw two_nsmul,
+  apply quotient.sound,
+  apply pgame.add_pow_half_succ_self_eq_pow_half,
+end
+
+/-- Exponentiation map from natural numbers to powers of integers. -/
+def pow (a : ℤ) (n : ℕ) : submonoid.powers a := ⟨a ^ n, n, rfl⟩
+
+/-- Logarithms from powers of integers to natural numbers. -/
+def log {a : ℤ} (p : submonoid.powers a) : ℕ :=
+nat.find $ (submonoid.mem_powers_iff a p.val).1 p.prop
+
+@[simp] theorem log_pow_eq_self (a : ℤ) (ha : 2 ≤ a) (n : ℕ) : log (pow a n) = n :=
+begin
+  unfold log,
+  generalize_proofs h,
+  exact @int.pow_right_injective a ha (nat.find h) n (nat.find_spec h),
+end
+
+@[simp] theorem pow_log_eq_self {a : ℤ} (n : submonoid.powers a) : pow a (log n) = n :=
+begin
+  unfold pow,
+  unfold log,
+  rcases n with ⟨_, hn⟩,
+  congr,
+  exact nat.find_spec hn,
+end
+
+lemma nsmul_pow_two_pow_half (n : ℕ) : 2 ^ n • pow_half n = 1 :=
+begin
+  induction n with n hn,
+  { simp only [nsmul_one, pow_half_zero, nat.cast_one, pow_zero] },
+  { rw [← hn, ← double_pow_half_succ_eq_pow_half n, smul_smul (2^n) 2 (pow_half n.succ),
+        mul_comm, pow_succ] }
+end
+
+lemma nsmul_pow_two_pow_half' (n k : ℕ) : 2 ^ n • pow_half (n + k) = pow_half k :=
+begin
+  induction k with k hk,
+  { simp [nsmul_pow_two_pow_half] },
+  { rw [← double_pow_half_succ_eq_pow_half (n + k), ← double_pow_half_succ_eq_pow_half k] at hk,
+    have : 2 ^ n • 2 • pow_half (n + k).succ = 2 • 2 ^ n • pow_half (n + k).succ,
+      by { apply smul_algebra_smul_comm },
+    rw this at hk,
+    exact (@gmul_cancel_of_pos 2 (2^n • pow_half (n + k).succ) (pow_half k.succ) (by norm_num)) hk }
+end
+
+lemma nsmul_int_pow_two_pow_half (m : ℤ) (n k : ℕ) :
+  (m * 2 ^ n) • pow_half (n + k) = m • pow_half k :=
+begin
+  rw mul_gsmul,
+  congr,
+  norm_cast,
```

Comment:
no change, but wow that this line is required

Reply:
Type casting is a mystery!